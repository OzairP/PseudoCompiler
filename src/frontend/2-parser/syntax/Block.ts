import { Production } from '../transitions'
import { Kind, Node } from './Node'
import { Statement } from './statement'

export class Block extends Node<Kind.Block> {
	constructor(public start: number, public width: number, public statements: Array<Statement>) {
		super(Kind.Block, Production.BLOCK, start, width)
	}
}
