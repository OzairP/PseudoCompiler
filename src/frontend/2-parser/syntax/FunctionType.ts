import { Production } from '../transitions'
import { Kind, Node } from './Node'
import { Parameter } from './Parameter'
import { Type } from './Type'

export class FunctionType extends Node<Kind.FunctionType> {
	constructor(start: number, width: number, public parameters: Array<Parameter>, public returnType: Type) {
		super(Kind.FunctionType, Production.TYPE, start, width)
	}
}
