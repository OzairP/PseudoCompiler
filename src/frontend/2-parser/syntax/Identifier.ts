import { Production } from '../transitions'
import { Kind, Node } from './Node'

export class Identifier extends Node<Kind.Identifier> {
	constructor(public start: number, public width: number, public name: string) {
		super(Kind.Identifier, Production.IDEN, start, width)
	}
}
