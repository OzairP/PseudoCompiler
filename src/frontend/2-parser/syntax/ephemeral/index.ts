export * from './EphemeralBinaryOperationExpressionTail'
export * from './EphemeralParamsList'
