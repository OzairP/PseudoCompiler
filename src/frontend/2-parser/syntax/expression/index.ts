export * from './Expression'
export * from './BinaryOperationExpression'
