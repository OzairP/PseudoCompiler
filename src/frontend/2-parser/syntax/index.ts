export * from './expression'
export * from './literal'
export * from './statement'

export * from './Identifier'
export * from './Program'
export * from './Type'
export * from './Block'
export * from './Node'
