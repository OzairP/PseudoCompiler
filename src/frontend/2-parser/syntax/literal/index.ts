export * from './Literal'
export * from './StringLiteral'
export * from './NumericLiteral'
export * from './FloatLiteral'
