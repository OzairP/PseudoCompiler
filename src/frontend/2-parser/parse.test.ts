import * as Lexer from '../1-lexer'
import * as Parser from './index'
import { Program } from './syntax'

const ast = (literals: TemplateStringsArray, ...placeholders: Array<string>): Program =>
	Parser.parse(Lexer.lex(String.raw(literals, ...placeholders)))

describe('parser', () => {
	describe('expression', () => {
		test('identifier', () => {
			expect(ast`a;`).toMatchSnapshot()
		})

		describe('literal', () => {
			test('string literal', () => {
				expect(ast`"hello";`).toMatchSnapshot()
			})

			test('numeric literal', () => {
				expect(ast`1;`).toMatchSnapshot()
			})

			test('float literal', () => {
				expect(ast`1.0;`).toMatchSnapshot()
			})
		})

		test('parenthesized', () => {
			expect(ast`(1);`).toMatchSnapshot()
		})

		describe('binary operation', () => {
			test('binary add', () => {
				expect(ast`1 + 1;`).toMatchSnapshot()
			})

			test('binary subtract', () => {
				expect(ast`1 - 1;`).toMatchSnapshot()
			})

			test('binary multiply', () => {
				expect(ast`1 * 1;`).toMatchSnapshot()
			})

			test('binary divide', () => {
				expect(ast`1 / 1;`).toMatchSnapshot()
			})

			test('binary exponentiation', () => {
				expect(ast`1 ** 1;`).toMatchSnapshot()
			})

			test('binary modulo', () => {
				expect(ast`1 % 1;`).toMatchSnapshot()
			})

			test('binary strict compare', () => {
				expect(ast`1 === 1;`).toMatchSnapshot()
			})

			test('binary equality compare', () => {
				expect(ast`1 == 1;`).toMatchSnapshot()
			})

			test('binary less than equal', () => {
				expect(ast`1 <= 1;`).toMatchSnapshot()
			})

			test('binary greater than equal', () => {
				expect(ast`1 >= 1;`).toMatchSnapshot()
			})

			test('binary less than', () => {
				expect(ast`1 < 1;`).toMatchSnapshot()
			})

			test('binary greater than', () => {
				expect(ast`1 > 1;`).toMatchSnapshot()
			})

			test('binary AND', () => {
				expect(ast`1 && 1;`).toMatchSnapshot()
			})

			test('binary OR', () => {
				expect(ast`1 || 1;`).toMatchSnapshot()
			})

			test('complex binary expression', () => {
				expect(ast`1 - 1 * 5 + 1 % 3 + (2 * 3);`).toMatchSnapshot()
			})
		})

		test('assignment', () => {
			expect(ast`a = "foo";`).toMatchSnapshot()
		})

		describe('call', () => {
			test('niladic', () => {
				expect(ast`print();`).toMatchSnapshot()
			})

			test('monadic', () => {
				expect(ast`print(0);`).toMatchSnapshot()
			})

			test('dyadic', () => {
				expect(ast`print(0, 0);`).toMatchSnapshot()
			})

			test('triadic', () => {
				expect(ast`print(0, 0, 0);`).toMatchSnapshot()
			})

			test('polyadic', () => {
				expect(ast`print(0, 0, 0, 0);`).toMatchSnapshot()
			})
		})

		describe('function', () => {
			test('niladic', () => {
				expect(ast`
					func (): i32 {
						0;
					};
				`).toMatchSnapshot()
			})

			describe('monadic', () => {
				test('immutable', () => {
					expect(ast`
						func (i32 a): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('mutable', () => {
					expect(ast`
						func (mut i32 a): i32 {
							0;
						};
					`).toMatchSnapshot()
				})
			})

			describe('dyadic', () => {
				test('immutable', () => {
					expect(ast`
						func (i32 a, i32 b): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('mutable', () => {
					expect(ast`
						func (mut i32 a, mut i32 b): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('1 mutable', () => {
					expect(ast`
						func (mut i32 a, i32 b): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('2 mutable', () => {
					expect(ast`
						func (i32 a, mut i32 b): i32 {
							0;
						};
					`).toMatchSnapshot()
				})
			})

			describe('triadic', () => {
				test('immutable', () => {
					expect(ast`
						func (i32 a, i32 b, i32 c): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('mutable', () => {
					expect(ast`
						func (mut i32 a, mut i32 b, mut i32 c): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('1 mutable', () => {
					expect(ast`
						func (mut i32 a, i32 b, i32 c): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('2 mutable', () => {
					expect(ast`
						func (i32 a, mut i32 b, i32 c): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('3 mutable', () => {
					expect(ast`
						func (i32 a, i32 b, mut i32 c): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('1-2 mutable', () => {
					expect(ast`
						func (mut i32 a, mut i32 b, i32 c): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('2-3 mutable', () => {
					expect(ast`
						func (i32 a, mut i32 b, mut i32 c): i32 {
							0;
						};
					`).toMatchSnapshot()
				})

				test('1-3 mutable', () => {
					expect(ast`
						func (mut i32 a, i32 b, mut i32 c): i32 {
							0;
						};
					`).toMatchSnapshot()
				})
			})

			describe('function type', () => {
				test('partially mutable dyadic', () => {
					expect(ast`
						func ((i32 a, mut i32 b): i32 a): i32 {
							0;
						};
					`).toMatchSnapshot()
				})
			})
		})
	})

	describe('statement', () => {
		describe('variable declaration', () => {
			test('immutable inferred', () => {
				expect(ast`let foo = 1;`).toMatchSnapshot()
			})

			test('mutable inferred', () => {
				expect(ast`let mut foo = 1;`).toMatchSnapshot()
			})

			test('immutable explicit', () => {
				expect(ast`let foo: i32 = 1;`).toMatchSnapshot()
			})

			test('mutable explicit', () => {
				expect(ast`let mut foo: i32 = 1;`).toMatchSnapshot()
			})
		})

		test('return', () => {
			expect(ast`return 1;`).toMatchSnapshot()
		})

		describe('conditional statement', () => {
			test('if', () => {
				expect(ast`if (0) { 0; }`).toMatchSnapshot()
			})

			test('if else', () => {
				expect(ast`
					if (0) { 0; }
					else { 0; }
				`).toMatchSnapshot()
			})

			test('if elif', () => {
				expect(ast`
					if (0) { 0; } 
					elif (0) { 0; }
				`).toMatchSnapshot()
			})

			test('if elif else', () => {
				expect(ast`
					if (0) { 0; } 
					elif (0) { 0; }
					else { 0; }
				`).toMatchSnapshot()
			})

			test('if elif elif else', () => {
				expect(ast`
					if (0) { 0; } 
					elif (0) { 0; }
					elif (0) { 0; }
					else { 0; }
				`).toMatchSnapshot()
			})
		})

		describe('function declaration statement', () => {
			test('niladic', () => {
				expect(ast`
					func z (): i32 {
						0;
					}
				`).toMatchSnapshot()
			})

			describe('monadic', () => {
				test('monadic', () => {
					expect(ast`
						func z (i32 a): i32 {
							0;
						}
					`).toMatchSnapshot()
				})

				test('mutable monadic', () => {
					expect(ast`
						func z (mut i32 a): i32 {
							0;
						}
					`).toMatchSnapshot()
				})
			})

			describe('dyadic', () => {
				test('immutable', () => {
					expect(ast`
						func z (i32 a, i32 b): i32 {
							0;
						}
					`).toMatchSnapshot()
				})

				test('mutable', () => {
					expect(ast`
						func z (mut i32 a, mut i32 b): i32 {
							0;
						}
					`).toMatchSnapshot()
				})

				test('1 mutable', () => {
					expect(ast`
						func z (mut i32 a, i32 b): i32 {
							0;
						}
					`).toMatchSnapshot()
				})
				test('2 mutable', () => {
					expect(ast`
						func z (i32 a, mut i32 b): i32 {
							0;
						}
					`).toMatchSnapshot()
				})
			})

			describe('triadic', () => {
				test('immutable', () => {
					expect(ast`
						func z (i32 a, i32 b, i32 c): i32 {
							0;
						}
					`).toMatchSnapshot()
				})

				test('mutable', () => {
					expect(ast`
						func z (mut i32 a, mut i32 b, mut i32 c): i32 {
							0;
						}
					`).toMatchSnapshot()
				})

				test('1 mutable', () => {
					expect(ast`
						func z (mut i32 a, i32 b, i32 c): i32 {
							0;
						}
					`).toMatchSnapshot()
				})

				test('2 mutable', () => {
					expect(ast`
						func z (i32 a, mut i32 b, i32 c): i32 {
							0;
						}
					`).toMatchSnapshot()
				})

				test('3 mutable', () => {
					expect(ast`
						func z (i32 a, i32 b, mut i32 c): i32 {
							0;
						}
					`).toMatchSnapshot()
				})

				test('1-2 mutable', () => {
					expect(ast`
						func z (mut i32 a, mut i32 b, i32 c): i32 {
							0;
						}
					`).toMatchSnapshot()
				})

				test('2-3 mutable', () => {
					expect(ast`
						func z (i32 a, mut i32 b, mut i32 c): i32 {
							0;
						}
					`).toMatchSnapshot()
				})

				test('1-3 mutable', () => {
					expect(ast`
						func z (mut i32 a, i32 b, mut i32 c): i32 {
							0;
						}
					`).toMatchSnapshot()
				})
			})

			describe('function type', () => {
				test('partially mutable dyadic', () => {
					expect(ast`
						func z ((i32 a, mut i32 b): i32 a): i32 {
							0;
						}
					`).toMatchSnapshot()
				})
			})
		})
	})

	describe('multi statement', () => {
		test('two', () => {
			expect(ast`0; 0;`).toMatchSnapshot()
		})

		test('three', () => {
			expect(ast`0; 0; 0;`).toMatchSnapshot()
		})
	})
})
