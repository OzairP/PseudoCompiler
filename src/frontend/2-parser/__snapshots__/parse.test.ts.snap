// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser expression binary operation binary AND 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 5,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 5,
      \\"operator\\": \\"and_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 5,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary OR 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 5,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 5,
      \\"operator\\": \\"or_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 5,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary add 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 4,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 4,
      \\"operator\\": \\"plus_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 4,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary divide 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 4,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 4,
      \\"operator\\": \\"divide_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 4,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary equality compare 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 5,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 5,
      \\"operator\\": \\"equality_cmp_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 5,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary exponentiation 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 5,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 5,
      \\"operator\\": \\"exponentiation_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 5,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary greater than 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 4,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 4,
      \\"operator\\": \\"greater_than_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 4,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary greater than equal 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 5,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 5,
      \\"operator\\": \\"greater_than_eql_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 5,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary less than 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 4,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 4,
      \\"operator\\": \\"less_than_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 4,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary less than equal 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 5,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 5,
      \\"operator\\": \\"less_than_eql_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 5,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary modulo 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 4,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 4,
      \\"operator\\": \\"modulo_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 4,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary multiply 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 4,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 4,
      \\"operator\\": \\"multiply_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 4,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary strict compare 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 6,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 6,
      \\"operator\\": \\"strict_cmp_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 6,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation binary subtract 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 4,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 4,
      \\"operator\\": \\"minus_op\\",
      \\"lhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 1,
        \\"value\\": 1
      },
      \\"rhs\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 4,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser expression binary operation complex binary expression 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 20,
  \\"statements\\": [
    {
      \\"kind\\": 15,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 20,
      \\"operator\\": \\"plus_op\\",
      \\"lhs\\": {
        \\"kind\\": 15,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 0,
        \\"width\\": 13,
        \\"operator\\": \\"modulo_op\\",
        \\"lhs\\": {
          \\"kind\\": 15,
          \\"token\\": \\"EXPR\\",
          \\"start\\": 0,
          \\"width\\": 10,
          \\"operator\\": \\"plus_op\\",
          \\"lhs\\": {
            \\"kind\\": 15,
            \\"token\\": \\"EXPR\\",
            \\"start\\": 0,
            \\"width\\": 7,
            \\"operator\\": \\"multiply_op\\",
            \\"lhs\\": {
              \\"kind\\": 15,
              \\"token\\": \\"EXPR\\",
              \\"start\\": 0,
              \\"width\\": 4,
              \\"operator\\": \\"minus_op\\",
              \\"lhs\\": {
                \\"kind\\": 18,
                \\"token\\": \\"EXPR\\",
                \\"start\\": 0,
                \\"width\\": 1,
                \\"value\\": 1
              },
              \\"rhs\\": {
                \\"kind\\": 18,
                \\"token\\": \\"EXPR\\",
                \\"start\\": 4,
                \\"width\\": 1,
                \\"value\\": 1
              }
            },
            \\"rhs\\": {
              \\"kind\\": 18,
              \\"token\\": \\"EXPR\\",
              \\"start\\": 8,
              \\"width\\": 1,
              \\"value\\": 5
            }
          },
          \\"rhs\\": {
            \\"kind\\": 18,
            \\"token\\": \\"EXPR\\",
            \\"start\\": 12,
            \\"width\\": 1,
            \\"value\\": 1
          }
        },
        \\"rhs\\": {
          \\"kind\\": 18,
          \\"token\\": \\"EXPR\\",
          \\"start\\": 16,
          \\"width\\": 1,
          \\"value\\": 3
        }
      },
      \\"rhs\\": {
        \\"kind\\": 15,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 21,
        \\"width\\": 4,
        \\"parenthesized\\": true,
        \\"operator\\": \\"multiply_op\\",
        \\"lhs\\": {
          \\"kind\\": 18,
          \\"token\\": \\"EXPR\\",
          \\"start\\": 21,
          \\"width\\": 1,
          \\"value\\": 2
        },
        \\"rhs\\": {
          \\"kind\\": 18,
          \\"token\\": \\"EXPR\\",
          \\"start\\": 25,
          \\"width\\": 1,
          \\"value\\": 3
        }
      }
    }
  ]
}"
`;

exports[`parser expression identifier 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 1,
  \\"statements\\": [
    {
      \\"kind\\": 3,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 1,
      \\"name\\": \\"a\\"
    }
  ]
}"
`;

exports[`parser expression literals float literal 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 3,
  \\"statements\\": [
    {
      \\"kind\\": 19,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 3,
      \\"value\\": 1
    }
  ]
}"
`;

exports[`parser expression literals numeric literal 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 1,
  \\"statements\\": [
    {
      \\"kind\\": 18,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 1,
      \\"value\\": 1
    }
  ]
}"
`;

exports[`parser expression literals string literal 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 7,
  \\"statements\\": [
    {
      \\"kind\\": 17,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 7,
      \\"value\\": \\"hello\\"
    }
  ]
}"
`;

exports[`parser expression parenthesized expression 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 1,
  \\"width\\": 1,
  \\"statements\\": [
    {
      \\"kind\\": 18,
      \\"token\\": \\"STMT\\",
      \\"start\\": 1,
      \\"width\\": 1,
      \\"parenthesized\\": true,
      \\"value\\": 1
    }
  ]
}"
`;

exports[`parser statement return 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 9,
  \\"statements\\": [
    {
      \\"kind\\": 9,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 9,
      \\"expression\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 7,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser statement variable declaration immutable explicit 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 17,
  \\"statements\\": [
    {
      \\"kind\\": 4,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 17,
      \\"mutable\\": false,
      \\"type\\": {
        \\"kind\\": 25,
        \\"token\\": \\"TYPE\\",
        \\"start\\": 9,
        \\"width\\": 3
      },
      \\"name\\": {
        \\"kind\\": 3,
        \\"token\\": \\"IDEN\\",
        \\"start\\": 4,
        \\"width\\": 3,
        \\"name\\": \\"foo\\"
      },
      \\"assignment\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 15,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser statement variable declaration immutable inferred 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 12,
  \\"statements\\": [
    {
      \\"kind\\": 4,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 12,
      \\"mutable\\": false,
      \\"name\\": {
        \\"kind\\": 3,
        \\"token\\": \\"IDEN\\",
        \\"start\\": 4,
        \\"width\\": 3,
        \\"name\\": \\"foo\\"
      },
      \\"assignment\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 10,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser statement variable declaration mutable explicit 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 21,
  \\"statements\\": [
    {
      \\"kind\\": 4,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 21,
      \\"mutable\\": true,
      \\"type\\": {
        \\"kind\\": 25,
        \\"token\\": \\"TYPE\\",
        \\"start\\": 13,
        \\"width\\": 3
      },
      \\"name\\": {
        \\"kind\\": 3,
        \\"token\\": \\"IDEN\\",
        \\"start\\": 8,
        \\"width\\": 3,
        \\"name\\": \\"foo\\"
      },
      \\"assignment\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 19,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;

exports[`parser statement variable declaration mutable inferred 1`] = `
"{
  \\"kind\\": 0,
  \\"token\\": \\"PROGRAM\\",
  \\"start\\": 0,
  \\"width\\": 16,
  \\"statements\\": [
    {
      \\"kind\\": 4,
      \\"token\\": \\"STMT\\",
      \\"start\\": 0,
      \\"width\\": 16,
      \\"mutable\\": true,
      \\"name\\": {
        \\"kind\\": 3,
        \\"token\\": \\"IDEN\\",
        \\"start\\": 8,
        \\"width\\": 3,
        \\"name\\": \\"foo\\"
      },
      \\"assignment\\": {
        \\"kind\\": 18,
        \\"token\\": \\"EXPR\\",
        \\"start\\": 14,
        \\"width\\": 1,
        \\"value\\": 1
      }
    }
  ]
}"
`;
